### ISSUE TITLE
Upgrade Mail to 2.7.x

### ISSUE THREAD

**Issue by garethrees** [mysociety/alaveteli#4896]
[2.7.0](https://rubygems.org/gems/mail/versions/2.7.0) is out.

We attempted in https://github.com/mysociety/alaveteli/pull/4422, but looks like there were some issues to fix.

Required by https://github.com/mysociety/alaveteli/issues/3907
Possibly required by https://github.com/mysociety/alaveteli/issues/3465

**Comment by garethrees** [comment]
Target release 0.35 #5229

**Comment by gbp** [comment]
Tested with #5452 - still the same issues as previously <details>
<summary>15 failures</summary>

```
$ rspec spec/mailers/application_mailer_spec.rb spec/mailers/request_mailer_spec.rb spec/script/handle-mail-replies_spec.rb spec/mailers/alaveteli_pro/subscription_mailer_spec.rb spec/lib/mail_handler/mail_handler_spec.rb spec/lib/alaveteli_mail_poller_spec.rb

Randomized with seed 9228
.....F....F..F...F...F.F..F...Non US-ASCII detected and no charset defined.
Defaulting to UTF-8, set your own if this is incorrect.
.........*...Non US-ASCII detected and no charset defined.
Defaulting to UTF-8, set your own if this is incorrect.
........................F...............................................F................F......F....F.F..F........F..................

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) when getting attachment attributes should ignore anything beyond the final MIME boundary
     # No reason given
     # ./spec/lib/mail_handler/mail_handler_spec.rb:432


Failures:

  1) when creating a mail object from raw data should not fail on invalid byte sequence in content-disposition header
     Failure/Error: part = Mail::Part.new("Content-Disposition: inline; filename=a\xB8z\r\n\r\nThis is the body text.")

     ArgumentError:
       invalid byte sequence in UTF-8
     # ./spec/lib/mail_handler/mail_handler_spec.rb:26:in `new'
     # ./spec/lib/mail_handler/mail_handler_spec.rb:26:in `block (2 levels) in <top (required)>'

  2) when creating a mail object from raw data should not be confused by subject lines with malformed UTF-8 at the end
     Failure/Error: expect(mail.subject).to eq('hello')

       expected: "hello"
            got: "hello�"

       (compared using ==)
     # ./spec/lib/mail_handler/mail_handler_spec.rb:52:in `block (2 levels) in <top (required)>'

  3) when getting the content type of a mail part should correctly return the types in an example bounce report
     Failure/Error: part.content_type ? part.content_type.split(';')[0] : nil

     NoMethodError:
       undefined method `content_type' for nil:NilClass
     # ./lib/mail_handler/backends/mail_backend.rb:156:in `get_content_type'
     # ./spec/lib/mail_handler/mail_handler_spec.rb:290:in `block (2 levels) in <top (required)>'

  4) ApplicationMailer when using plugins when a plugin appends its mail templates to the view paths should render a multipart email using a core template
     Failure/Error: expect(message_part).to match("Core multipart")

       expected "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-Di...r\n filename=original.eml\r\nContent-ID: <5dc41116cac40_fe823fc08003202021764@aion.mail>\r\n\r\nxxx" to match "Core multipart"
       Diff:
       @@ -1,2 +1,9 @@
       -Core multipart
       +Content-Type: message/rfc822;
       + filename=original.eml
       +Content-Transfer-Encoding:
       +Content-Disposition: attachment;
       + filename=original.eml
       +Content-ID: <5dc41116cac40_fe823fc08003202021764@aion.mail>
       +
       +xxx

     # ./spec/mailers/application_mailer_spec.rb:141:in `block (4 levels) in <top (required)>'

  5) ApplicationMailer when using plugins when a plugin appends its mail templates to the view paths should render a multipart email using a theme template
     Failure/Error: expect(message_part).to match("Theme multipart")

       expected "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-Di...r\n filename=original.eml\r\nContent-ID: <5dc41116d9409_fe823fc080032020219ed@aion.mail>\r\n\r\nxxx" to match "Theme multipart"
       Diff:
       @@ -1,2 +1,9 @@
       -Theme multipart
       +Content-Type: message/rfc822;
       + filename=original.eml
       +Content-Transfer-Encoding:
       +Content-Disposition: attachment;
       + filename=original.eml
       +Content-ID: <5dc41116d9409_fe823fc080032020219ed@aion.mail>
       +
       +xxx

     # ./spec/mailers/application_mailer_spec.rb:150:in `block (4 levels) in <top (required)>'

  6) ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a theme template
     Failure/Error: expect(message_part).to match("Theme multipart")

       expected "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-Di...r\n filename=original.eml\r\nContent-ID: <5dc41116e0c15_fe823fc080032020221be@aion.mail>\r\n\r\nxxx" to match "Theme multipart"
       Diff:
       @@ -1,2 +1,9 @@
       -Theme multipart
       +Content-Type: message/rfc822;
       + filename=original.eml
       +Content-Transfer-Encoding:
       +Content-Disposition: attachment;
       + filename=original.eml
       +Content-ID: <5dc41116e0c15_fe823fc080032020221be@aion.mail>
       +
       +xxx

     # ./spec/mailers/application_mailer_spec.rb:94:in `block (4 levels) in <top (required)>'

  7) ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a core template
     Failure/Error: expect(message_part).to match("Core multipart")

       expected "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-Di...r\n filename=original.eml\r\nContent-ID: <5dc41116e9f8c_fe823fc0800320202236b@aion.mail>\r\n\r\nxxx" to match "Core multipart"
       Diff:
       @@ -1,2 +1,9 @@
       -Core multipart
       +Content-Type: message/rfc822;
       + filename=original.eml
       +Content-Transfer-Encoding:
       +Content-Disposition: attachment;
       + filename=original.eml
       +Content-ID: <5dc41116e9f8c_fe823fc0800320202236b@aion.mail>
       +
       +xxx

     # ./spec/mailers/application_mailer_spec.rb:103:in `block (4 levels) in <top (required)>'

  8) RequestMailer when sending a new response email should send normal users a direct link
     Failure/Error: expect(mail_url).to eq expected_url

       expected: "http://test.host/en/request/here_is_a_character_that_needs_q?nocache=incoming-15#incoming-15"
            got: "http://test.host/en/request/here_is_a_character_that_needs_q?nocache=incoming-15#incoming-15\r"

       (compared using ==)
     # ./spec/mailers/request_mailer_spec.rb:492:in `block (3 levels) in <top (required)>'

  9) AlaveteliPro::SubscriptionMailer.payment_failed renders the body correctly
     Failure/Error:
       expect(subject.body.to_s).
         to eq(read_described_class_fixture('payment_failed'))

       expected: "Dear Paul Pro,\n\nThank you for your interest in Alaveteli Professional.\n\nWe’ve tried to bill your...bscriptions page: http://test.host/en/profile/subscriptions\n\n-- the Alaveteli Professional team\n"
            got: "Dear Paul Pro,\r\n\r\nThank you for your interest in Alaveteli Professional.\r\n\r\nWe’ve tried to b...tions page: http://test.host/en/profile/subscriptions\r\n\r\n-- the Alaveteli Professional team\r\n"

       (compared using ==)

       Diff:

     # ./spec/mailers/alaveteli_pro/subscription_mailer_spec.rb:19:in `block (3 levels) in <top (required)>'

  10) When filtering should detect a WebShield delivery error message
      Failure/Error: expect(r.status).to eq(1)

        expected: 1
             got: 0

        (compared using ==)
      # ./spec/script/handle-mail-replies_spec.rb:42:in `block (2 levels) in <top (required)>'

  11) AlaveteliMailPoller#poll_for_incoming if there is a timeout connecting to the POP server sends an exception notification
      Failure/Error:
        expect(notification.subject).
          to eq('[ERROR] (Timeout::Error) "execution expired"')

        expected: "[ERROR] (Timeout::Error) \"execution expired\""
             got: "[ERROR]  (Timeout::Error) \"execution expired\""

        (compared using ==)
      # ./spec/lib/alaveteli_mail_poller_spec.rb:275:in `block (4 levels) in <top (required)>'

  12) AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error receiving the mail sends an exception notification
      Failure/Error:
        expect(notification.subject).
          to eq('[ERROR] (ActiveRecord::StatementInvalid) "Deadlock"')

        expected: "[ERROR] (ActiveRecord::StatementInvalid) \"Deadlock\""
             got: "[ERROR]  (ActiveRecord::StatementInvalid) \"Deadlock\""

        (compared using ==)
      # ./spec/lib/alaveteli_mail_poller_spec.rb:154:in `block (5 levels) in <top (required)>'

  13) AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error getting the text of a mail sends an exception notification
      Failure/Error:
        expect(notification.subject).
          to eq('[ERROR] (Net::POPError) "Error code"')

        expected: "[ERROR] (Net::POPError) \"Error code\""
             got: "[ERROR]  (Net::POPError) \"Error code\""

        (compared using ==)
      # ./spec/lib/alaveteli_mail_poller_spec.rb:90:in `block (5 levels) in <top (required)>'

  14) AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error getting the unique ID of a mail sends an exception notification
      Failure/Error:
        expect(notification.subject).
          to eq('[ERROR] (Net::POPError) "Error code"')

        expected: "[ERROR] (Net::POPError) \"Error code\""
             got: "[ERROR]  (Net::POPError) \"Error code\""

        (compared using ==)
      # ./spec/lib/alaveteli_mail_poller_spec.rb:73:in `block (5 levels) in <top (required)>'

  15) AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error deleting the mail sends an exception notification
      Failure/Error:
        expect(exception_notification.subject).
          to eq('[ERROR] (Net::POPError) "Error code"')

        expected: "[ERROR] (Net::POPError) \"Error code\""
             got: "[ERROR]  (Net::POPError) \"Error code\""

        (compared using ==)
      # ./spec/lib/alaveteli_mail_poller_spec.rb:191:in `block (5 levels) in <top (required)>'

Finished in 1 minute 2.96 seconds (files took 8.63 seconds to load)
177 examples, 15 failures, 1 pending

Failed examples:

rspec ./spec/lib/mail_handler/mail_handler_spec.rb:25 # when creating a mail object from raw data should not fail on invalid byte sequence in content-disposition header
rspec ./spec/lib/mail_handler/mail_handler_spec.rb:47 # when creating a mail object from raw data should not be confused by subject lines with malformed UTF-8 at the end
rspec ./spec/lib/mail_handler/mail_handler_spec.rb:286 # when getting the content type of a mail part should correctly return the types in an example bounce report
rspec ./spec/mailers/application_mailer_spec.rb:135 # ApplicationMailer when using plugins when a plugin appends its mail templates to the view paths should render a multipart email using a core template
rspec ./spec/mailers/application_mailer_spec.rb:144 # ApplicationMailer when using plugins when a plugin appends its mail templates to the view paths should render a multipart email using a theme template
rspec ./spec/mailers/application_mailer_spec.rb:88 # ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a theme template
rspec ./spec/mailers/application_mailer_spec.rb:97 # ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a core template
rspec ./spec/mailers/request_mailer_spec.rb:487 # RequestMailer when sending a new response email should send normal users a direct link
rspec ./spec/mailers/alaveteli_pro/subscription_mailer_spec.rb:18 # AlaveteliPro::SubscriptionMailer.payment_failed renders the body correctly
rspec ./spec/script/handle-mail-replies_spec.rb:40 # When filtering should detect a WebShield delivery error message
rspec ./spec/lib/alaveteli_mail_poller_spec.rb:272 # AlaveteliMailPoller#poll_for_incoming if there is a timeout connecting to the POP server sends an exception notification
rspec ./spec/lib/alaveteli_mail_poller_spec.rb:151 # AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error receiving the mail sends an exception notification
rspec ./spec/lib/alaveteli_mail_poller_spec.rb:87 # AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error getting the text of a mail sends an exception notification
rspec ./spec/lib/alaveteli_mail_poller_spec.rb:70 # AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error getting the unique ID of a mail sends an exception notification
rspec ./spec/lib/alaveteli_mail_poller_spec.rb:188 # AlaveteliMailPoller#poll_for_incoming if there is mail on the POP server if there is an error deleting the mail sends an exception notification

Randomized with seed 9228
```

</details>

Going to bump to 0.37 so this doesn't block 0.36

**Comment by garethrees** [comment]
Requires #5159

**Comment by gbp** [comment]
So there are [15 spec failures](https://travis-ci.org/github/mysociety/alaveteli/jobs/710507486) I've gone through them and worked out which commit upstream resulted in our specs failing.

### https://github.com/mikel/mail/commit/5bcf5df9b4e9c3b737fbe2b67b1d236be999b153
- `rspec spec/lib/alaveteli_mail_poller_spec.rb:151` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/alaveteli_mail_poller_spec.rb#L151)
- `rspec spec/lib/alaveteli_mail_poller_spec.rb:188` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/alaveteli_mail_poller_spec.rb#L188)
- `rspec spec/lib/alaveteli_mail_poller_spec.rb:272` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/alaveteli_mail_poller_spec.rb#L272)
- `rspec spec/lib/alaveteli_mail_poller_spec.rb:70` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/alaveteli_mail_poller_spec.rb#L70)
- `rspec spec/lib/alaveteli_mail_poller_spec.rb:87` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/alaveteli_mail_poller_spec.rb#L87)

### https://github.com/mikel/mail/commit/d17be66765f7f63caa5d5cedb67a597f853a4402
- `rspec spec/lib/mail_handler/mail_handler_spec.rb:25` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/mail_handler/mail_handler_spec.rb#L25)

### https://github.com/mikel/mail/commit/7996f06f27e73e76f644041f79853b37e7b022a3
- `rspec spec/lib/mail_handler/mail_handler_spec.rb:286` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/mail_handler/mail_handler_spec.rb#L286)
- `rspec spec/script/handle-mail-replies_spec.rb:40` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/script/handle-mail-replies_spec.rb#L40) also had to comment out [an earlier spec](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/script/handle-mail-replies_spec.rb#L10) due to a gemspec deprecation warning being outputted (relating to `#has_rdoc` option) to get the same failure as on CI.

### https://github.com/mikel/mail/commit/8e15ff55e9162f5f9b9dd0038c543a567f261156
- `rspec spec/lib/mail_handler/mail_handler_spec.rb:47` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/lib/mail_handler/mail_handler_spec.rb#L47)

### https://github.com/mikel/mail/commit/6a019d5372c3792f32ab8423fd65f477e783856a
- `rspec spec/mailers/alaveteli_pro/subscription_mailer_spec.rb:18` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/alaveteli_pro/subscription_mailer_spec.rb#L18)

### https://github.com/mikel/mail/commit/efc735f19c2b6a17f2eb5e4f80d9da6a18f8cfc1
- `rspec spec/mailers/application_mailer_spec.rb:135` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/application_mailer_spec.rb#L135)
- `rspec spec/mailers/application_mailer_spec.rb:144` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/application_mailer_spec.rb#L144)
- `rspec spec/mailers/application_mailer_spec.rb:88` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/application_mailer_spec.rb#L88)
- `rspec spec/mailers/application_mailer_spec.rb:97` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/application_mailer_spec.rb#L97)

### https://github.com/mikel/mail/commit/6a019d5372c3792f32ab8423fd65f477e783856a
- `rspec spec/mailers/request_mailer_spec.rb:487` [*](https://github.com/mysociety/alaveteli/blob/f0004a9/spec/mailers/request_mailer_spec.rb#L487)

**Comment by gbp** [comment]
Working through fixing these:

- [x] https://github.com/mikel/mail/commit/5bcf5df9b4e9c3b737fbe2b67b1d236be999b153 - fixed by 78b24c0f1a
- [ ] https://github.com/mikel/mail/commit/d17be66765f7f63caa5d5cedb67a597f853a4402
- [ ] https://github.com/mikel/mail/commit/7996f06f27e73e76f644041f79853b37e7b022a3
- [ ] https://github.com/mikel/mail/commit/8e15ff55e9162f5f9b9dd0038c543a567f261156
- [ ] https://github.com/mikel/mail/commit/6a019d5372c3792f32ab8423fd65f477e783856a
- [x] https://github.com/mikel/mail/commit/efc735f19c2b6a17f2eb5e4f80d9da6a18f8cfc1 - fixed by 947382d493
